File: src/main/java/no/nav/sbl/service/EnheterCache.java
package no.nav.sbl.service;

import lombok.extern.slf4j.Slf4j;
import no.nav.sbl.consumers.norg2.Norg2Client;
import no.nav.sbl.consumers.norg2.domain.Norg2EnheterResponse;
import no.nav.sbl.rest.domain.DecoratorDomain;
import org.springframework.scheduling.annotation.Scheduled;

import javax.inject.Inject;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

import static java.util.Collections.unmodifiableList;
import static java.util.Collections.unmodifiableMap;

@Slf4j
public class EnheterCache {
    private static final long HVER_TOLVTE_TIME = 12 * 3600 * 1000;
    private Map<String, DecoratorDomain.Enhet> cache = unmodifiableMap(new HashMap<>());
    private List<DecoratorDomain.Enhet> cacheList = unmodifiableList(new ArrayList<>());

    @Inject
    Norg2Client norg2Client;

    @Scheduled(fixedRate = HVER_TOLVTE_TIME)
    private void refreshCache() {
        try {
            Norg2EnheterResponse response = norg2Client.hentAlleEnheter();

            cacheList = unmodifiableList(response.enheter
                    .stream()
                    .map((enhet) -> new DecoratorDomain.Enhet(enhet.getEnhetNr(), enhet.getNavn()))
                    .sorted(Comparator.comparing(DecoratorDomain.Enhet::getEnhetId))
                    .collect(Collectors.toList())
            );

            cache = unmodifiableMap(cacheList
                    .stream()
                    .collect(Collectors.toMap(
                            enhet -> enhet.enhetId,
                            Function.identity()
                    )));

        } catch (Exception e) {
            log.error("Kunne ikke hente ut alle aktive enheter fra NORG2-rest", e);
        }
    }

    public Map<String, DecoratorDomain.Enhet> get() {
        return this.cache;
    }

    public List<DecoratorDomain.Enhet> getAll() {
        return cacheList;
    }
}

File: src/main/java/no/nav/sbl/service/EnheterCache.kt
package no.nav.sbl.service

import no.nav.sbl.consumers.norg2.Norg2Client
import no.nav.sbl.consumers.norg2.domain.Enhet
import no.nav.sbl.rest.domain.DecoratorDomain
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.scheduling.annotation.Scheduled
import java.util.*

class EnheterCache(
    @Autowired private val norg2Client: Norg2Client
) {
    companion object {
        private const val HVER_TOLVTE_TIME: Long = 12 * 3600 * 1000
    }

    private val log = LoggerFactory.getLogger(EnheterCache::class.java)
    private var cache: Map<String, DecoratorDomain.Enhet> = Collections.unmodifiableMap(HashMap())
    private var cacheList: List<DecoratorDomain.Enhet> = Collections.unmodifiableList(ArrayList())

    @Scheduled(fixedRate = HVER_TOLVTE_TIME)
    private fun refreshCache() {
        try {
            val enheter: List<Enhet> = norg2Client.hentAlleEnheter()

            cacheList = Collections.unmodifiableList(enheter
                .map { enhet -> DecoratorDomain.Enhet(enhet.enhetNr, enhet.navn) }
                .sortedBy { it.enhetId }
            )

            cache = Collections.unmodifiableMap(cacheList
                .associateBy { it.enhetId }
            )

        } catch (e: Exception) {
            log.error("Kunne ikke hente ut alle aktive enheter fra NORG2-rest", e)
        }
    }

    fun get(): Map<String, DecoratorDomain.Enhet> {
        return this.cache
    }

    fun getAll(): List<DecoratorDomain.Enhet> {
        return cacheList
    }
}